boofuzz/blocks/block.py

```diff
@@ -38,254 +46,141 @@ def __init__(
+    #                 break
+    #
+    #         # if the possible mutations for the stack are exhausted.
+    #         if not mutated:
+    #             # increment the group value index.
+    #             self.group_idx += 1
+    #
+    #             # if the group values are exhausted, we are done with this block.
+    #             if self.group_idx == group_count:
+    #                 # restore the original group value.
+    #                 self.request.names[self.group].reset()
+    #
+    #             # otherwise continue mutating this group/block.
+    #             else:
+    #                 # update the group value to that at the current index.
+    #                 self.request.names[self.group]._value = self.request.names[self.group].values[self.group_idx]
+    #
+    #                 # this the mutate state for every item in this blocks stack.
+    #                 # NOT THE BLOCK ITSELF THOUGH! (hence why we didn't call self.reset())
+    #                 for item in self.stack:
+    #                     if item.fuzzable:
+    #                         item.reset()
+    #
+    #                 # now mutate the first field in this block before continuing.
+    #                 # (we repeat a test case if we don't mutate something)
+    #                 for item in self.stack:
+    #                     if item.fuzzable and item.mutate():
+    #                         mutated = True
+    #
+    #                         if not isinstance(item, Block):
+    #                             self.request.mutant = item
+    #
+    #                         break
+    #     #
+    #     # no grouping, mutate every item on the stack once.
+    #     #
+    #     else:
+    #         for item in self.stack:
+    #             if item.fuzzable and item.mutate():
+    #                 mutated = True
+    #
+    #                 if not isinstance(item, Block):
+    #                     self.request.mutant = item
+    #
+    #                 break
+    #
+    #     # if this block is dependant on another field, then manually update that fields value appropriately while we
+    #     # mutate this block. we'll restore the original value of the field prior to continuing.
+    #     if mutated and self.dep:
+    #         # if a list of values was specified, use the first item in the list.
+    #         if self.dep_values:
+    #             self.request.names[self.dep]._value = self.dep_values[0]
+    #
+    #         # if a list of values was not specified, assume a single value is present.
+    #         else:
+    #             self.request.names[self.dep]._value = self.dep_value
+    #
+    #     # we are done mutating this block.
+    #     if not mutated:
+    #         self._fuzz_complete = True
+    #
+    #         # if we had a dependency, make sure we restore the original value.
+    #         if self.dep:
+    #             self.request.names[self.dep].reset()
+    #
+    #     return mutated
```

Comment:
Is this todo still relevant?
Can we remove the orphaned code?

Reply:
Good catch -- for full backwards compatibility, we need to port this to enable the behavior where a block will do all its mutations for each group value. Interestingly, this is a special case of mutating multiple items at one time.

To do the port, we'd add a `def mutations(self, default_value):` to override the FuzzableBlock mutations code.

Reply:
Can you give an example of how this would be used?

Reply:
I think we're talking about this
https://github.com/jtpereyda/boofuzz/blob/898a1d2a1927ef22128f93de5d33bcbc39075b47/request_definitions/http.py#L15-L16

Have the block iterate over all group values while fuzzing the contents of the block.

Reply:
Currently, this branch will always use `GET` and and iterate over all child primitives, while it should repeat fuzzing with `HEAD`, ...

Reply:
I'm trying to reimplement this feature but I can't find a way to mutate over multiple item (the group and item stack) at the same time.

Any ideas?

Reply:
@jtpereyda do you have a hint on how to implement the group mutations feature?
Apart from the name resolving discussion this is the only thing holding us back from merging.

Reply:
@SR4ven I didn't realize how difficult this bit would be! The old implementation manually mucks with the Group primitive's values. The new implementation doesn't quite work that way. This is actually an instance where we want to mutate two primitives at one time. Great time to try it out!

In Fuzzable.get_mutations, see this excerpt:
```
            for value in itertools.chain(self.mutations(self.original_value()), self._fuzz_values):
                if self._halt_mutations:
                    self._halt_mutations = False
                    return
                if isinstance(value, Mutation):
                    yield value
                else:
                    yield Mutation(mutations={self.qualified_name: value})
```
Especially the last 4 lines --  the mutations method can either render a value (a bytes object), or a Mutation object. The Mutation.muations field is a map from qualified names to mutated values. get_mutations will always normalize the mutation into a `Mutation` object.

So, you can in Block create a Mutation object, add the Group mutation (`"qualified_group_name": b"the_value"`), then manually add the individual mutation based on what you iterated from the item stack.

Pseudocode:
```
for value in theGroup.mutations():
    m = Mutation(mutations={self.request.resolve_name(self.context_path, self.group): value})  # maybe use self.group.context_path?
    for item in self.stack:
        self.request.mutant = item  # not sure about this -- the "mutant" field starts to become meaningless and may need to be reworked or removed eventually
        for mutation in item.get_mutations():
            # merge mutation.mutations into m.mutations
            yield mutation
```

The new design should in theory support and render these two mutations properly, but there may be bugs to work out.

Reply:
```python
def mutations(self, default_value):
    for value in self.request.resolve_name(self.context_path, self.group).mutations(default_value):
        m = Mutation(mutations={
            self.request.resolve_name(self.context_path, self.group): value})  # maybe use self.group.context_path?
        for item in self.stack:
            self.request.mutant = item  # not sure about this -- the "mutant" field starts to become meaningless and may need to be reworked or removed eventually
            for mutation in item.get_mutations():
                # merge mutation.mutations into m.mutations
                yield m
```
I tried this code in `Block` with the following request but it always yields `GETa`.
```python
    s_initialize("HTTP VERBS")
    with s_block("body", group="verbs"):
        s_group("verbs", values=["GET", "HEAD"])
        s_static("a")
```
`value` changes correctly, so maybe there's something wrong with the Mutation implementation?

Also, this example yields 4 test cases, though there are only 2 unique ones.

Reply:
Could use your help here @jtpereyda. Do you have time to look into it?

Reply:
Should be able to this week or next. Just had some priorities shift around enough to get back on this. I've been bummed to have left the half-done PR here for so long.

Reply:
I've been playing around with the groups and it seems to work fine with the examples.

However, the HTTP VERBS request example from above now yields `HEADa` twice now. At least it's not 4 test cases anymore but something still seems to go wrong.
Is that maybe because `GETa` is considered as the "good" case which is skipped but num_mutations is 2 so it yields the same values twice because the mutations are exhausted? In that case this request definition should only yield 1 test case right?

Reply:
Most likely caused by the group being inside the block itself. Since the rule is to mutate the block once for each mutation of the relevant group... but in this case that group is in the block... I can't quite wrap my head around what it _should_ do. I think if you take the verbs field outside of the block and put something else in, the behavior will make more sense.

I think we can leave it for now. If there are any ideas on the "right" answer when the group is a part of the block, feel free to record them here for future reference.